BDArmory/Control/BDModuleSurfaceAI.cs

```diff
@@ -390,27 +406,69 @@ void PilotLogic()
                                     continue;
                             }
                             dodgeVector = PredictCollisionWithVessel(vs.Current, 5f * predictMult, 0.5f);
-                            if (dodgeVector != null) break;
+                            if (dodgeVector != Vector3.zero) break;
                         }
+                    if (dodgeVector == Vector3.zero)
+                    {
+                        Vector3 vesselDir = (vessel.srfSpeed > 1 && !wasReversing) ? vessel.srf_vel_direction : vessel.vesselTransform.up;
+                        UnityEngine.Ray ray = new UnityEngine.Ray(vessel.CoM, vesselDir);
+                        // For most terrain, the spherecast produces a single hit, but for buildings and special scenery (e.g., Kerbal Konstructs with multiple colliders), multiple hits are detected.
+                        int hitCount = Physics.SphereCastNonAlloc(ray, terrainAlertDetectionRadius, terrainAvoidanceHits, terrainAlertThreatRange, (int)LayerMasks.Scenery); 
+                        if (hitCount == terrainAvoidanceHits.Length)
+                        {
+                            terrainAvoidanceHits = Physics.SphereCastAll(ray, terrainAlertDetectionRadius, terrainAlertThreatRange, (int)LayerMasks.Scenery);
+                            hitCount = terrainAvoidanceHits.Length;
+                        }
+                        if (hitCount > 0) // Found something. 
+                        {
+                            using (var hits = terrainAvoidanceHits.Take(hitCount).GetEnumerator())
+                                while (hits.MoveNext())
```

Comment:
Since there's no `break` in this, only the last hit (which isn't in any particular order) is going to do anything.

The way the pilot AI avoids sudden direction changes from buttresses and such in `FlyAvoidTerrain` is to average the hit normals weighted by distance (`alertNormal += normal / (1 + alertDistance * alertDistance);`), maybe something like that would work here?

The "heading into a wall" and `collisionTicker` stuff at the bottom should also be outside the `while` loop.

Reply:
alertNormal method seems to work.